controllers/virtualmachine/v1alpha2/virtualmachine_controller.go

```diff
@@ -0,0 +1,345 @@
+		if network == nil || (network.PrimaryIP4 == "" && network.PrimaryIP6 == "") {
+			return 10 * time.Second
+		}
+	}
+
+	return 0
+}
+
+func (r *Reconciler) ReconcileDelete(ctx *context.VirtualMachineContextA2) (reterr error) {
+	ctx.Logger.Info("Reconciling VirtualMachine Deletion")
+
+	if controllerutil.ContainsFinalizer(ctx.VM, finalizerName) {
+		defer func() {
+			r.Recorder.EmitEvent(ctx.VM, "Delete", reterr, false)
+		}()
+
+		if err := r.VMProvider.DeleteVirtualMachine(ctx, ctx.VM); err != nil {
+			ctx.Logger.Error(err, "Failed to delete VirtualMachine")
+			return err
+		}
+
+		controllerutil.RemoveFinalizer(ctx.VM, finalizerName)
+		ctx.Logger.Info("Provider Completed deleting Virtual Machine", "time", time.Now().Format(time.RFC3339))
+	}
+
+	// BMV: Shouldn't these be in the ContainsFinalizer block?
+	r.vmMetrics.DeleteMetrics(ctx)
+	r.Prober.RemoveFromProberManager(ctx.VM)
+
+	ctx.Logger.Info("Finished Reconciling VirtualMachine Deletion")
+	return nil
+}
+
+// ReconcileNormal processes a level trigger for this VM: create if it doesn't exist otherwise update the existing VM.
+func (r *Reconciler) ReconcileNormal(ctx *context.VirtualMachineContextA2) (reterr error) {
+	if !controllerutil.ContainsFinalizer(ctx.VM, finalizerName) {
+		// The finalizer must be present before proceeding in order to ensure that the VM will
+		// be cleaned up. Return immediately after here to let the patcher helper update the
+		// object, and then we'll proceed on the next reconciliation.
+		controllerutil.AddFinalizer(ctx.VM, finalizerName)
+		return nil
+	}
+
+	ctx.Logger.Info("Reconciling VirtualMachine")
+
+	defer func(beforeVMStatus *vmopv1.VirtualMachineStatus) {
+		// Log the reconcile time using the CR creation time and the time the VM reached the desired state
+		if reterr == nil && !apiequality.Semantic.DeepEqual(beforeVMStatus, &ctx.VM.Status) {
+			ctx.Logger.Info("Finished Reconciling VirtualMachine with updates to the CR",
+				"createdTime", ctx.VM.CreationTimestamp, "currentTime", time.Now().Format(time.RFC3339),
+				"spec.PowerState", ctx.VM.Spec.PowerState, "status.PowerState", ctx.VM.Status.PowerState)
+		} else {
+			ctx.Logger.Info("Finished Reconciling VirtualMachine")
+		}
+
+		beforeNoIP := beforeVMStatus.Network == nil || (beforeVMStatus.Network.PrimaryIP4 == "" && beforeVMStatus.Network.PrimaryIP6 == "")
+		nowWithIP := ctx.VM.Status.Network != nil && (ctx.VM.Status.Network.PrimaryIP4 != "" || ctx.VM.Status.Network.PrimaryIP6 != "")
+		if beforeNoIP && nowWithIP {
+			ctx.Logger.Info("VM successfully got assigned with an IP address", "time", time.Now().Format(time.RFC3339))
+		}
+	}(ctx.VM.Status.DeepCopy())
+
+	defer func() {
+		r.vmMetrics.RegisterVMCreateOrUpdateMetrics(ctx)
+	}()
+
+	if err := r.VMProvider.CreateOrUpdateVirtualMachine(ctx, ctx.VM); err != nil {
+		ctx.Logger.Error(err, "Failed to reconcile VirtualMachine")
+		r.Recorder.EmitEvent(ctx.VM, "CreateOrUpdate", err, false)
+		return err
+	}
+
```

Comment:
I'm doing that in the v1a2 provider now. I found it confusing that in v1a1 the responsibility of this field was share/duplicated between the controller & provider. IIRC I think v1a1 has a long standing bug too that it can set the Phase=Created when the VM hasn't actually be created.